### PR TITLE
aggfuncs: partially implement "AVG"

### PR DESCRIPTION
**Pull request by zz-jason**
## What have you changed? (mandatory)

partially implement the `AVG` function under the new aggregate function framework, only implemented the following `AVG` functions:
- AVG(DECIMAL)
- AVG(FLOAT64)

the following `AVG` functions are not implemented:
- AVG(DISTINCT DECIMAL)
- AVG(DISTINCT FLOAT64)

## What are the type of the changes (mandatory)?

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested (mandatory)?

- unit test
- explain test

## Does this PR affect documentation (docs/docs-cn) update? (optional)

No

## Refer to a related PR or issue link (optional)
To #6952. This is the partially supported `AVG` functions I mentioned in https://github.com/pingcap/tidb/pull/6852

### CODE REVIEW

File: executor/aggfuncs/aggfuncs.go [hunk 1]
```diff
@@ -23,9 +23,18 @@ import (
 
 // All the AggFunc implementations are listed here for navigation.
 var (
-// All the AggFunc implementations for "COUNT" are listed here.
-// All the AggFunc implementations for "SUM" are listed here.
-// All the AggFunc implementations for "AVG" are listed here.
+	// All the AggFunc implementations for "COUNT" are listed here.
+	// All the AggFunc implementations for "SUM" are listed here.
+	// All the AggFunc implementations for "AVG" are listed here.
+	_ AggFunc = (*avgOriginal4Decimal)(nil)
```
Comment (shenli) [inline executor/aggfuncs/aggfuncs.go]: Why name it Original?

File: executor/aggfuncs/func_avg.go [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+// All the AggFunc implementations for "AVG" are listed here.
+var (
+	_ AggFunc = (*avgOriginal4Decimal)(nil)
```
Comment (shenli) [inline executor/aggfuncs/func_avg.go]: Duplicate with the code in aggfuncs.go.

File: executor/aggfuncs/aggfuncs.go (hunk 1 above)
Comment (zz-jason) [reply executor/aggfuncs/aggfuncs.go]: means this aggregate function is in the `partial 1` or `complete` mode and it's input is the data from the child of logical aggregate operator, not the partial result of other aggregate function.

File: executor/aggfuncs/aggfuncs.go [hunk 3]
```diff
@@ -23,9 +23,18 @@ import (
...
+	_ AggFunc = (*avgPartial4Decimal)(nil)
+
+	_ AggFunc = (*avgOriginal4Float64)(nil)
+	_ AggFunc = (*avgPartial4Float64)(nil)
+
+	_ AggFunc = (*avgOriginal4Float32)(nil)
```
[the remark this answers is not in the document]
Comment (zz-jason) [reply executor/aggfuncs/aggfuncs.go]: Yes. But in order to be compatible with mysql, we can not wrap a `cast as real` when the input parameter is a `column` and its type is `float32`

[the remark this answers is not in the document]
Comment (XuHuaiyu) [reply executor/aggfuncs/aggfuncs.go]: avg(float) returns double in MySQL.

**Comment by zz-jason** [comment]
@XuHuaiyu @shenli PTAL

File: executor/aggfuncs/aggfuncs.go (hunk 1 above)
Comment (shenli) [reply executor/aggfuncs/aggfuncs.go]: How about `avg4Decimal`?

File: executor/aggfuncs/builder.go [hunk 4]
```diff
@@ -58,6 +59,42 @@ func buildSum(aggFuncDesc *aggregation.AggFuncDesc, ordinal int) AggFunc {
...
+	// duplicated input of the same group.
+	case aggregation.DedupMode:
+		return nil // not implemented yet.
+
+	// Build avg functions which consume the original data and update their
+	// partial results.
+	case aggregation.CompleteMode, aggregation.Partial1Mode:
+		switch aggFuncDesc.Args[0].GetType().Tp {
+		case mysql.TypeNewDecimal:
+			if aggFuncDesc.HasDistinct {
```
Comment (shenli) [inline executor/aggfuncs/builder.go]: Can we check it at the beginning of this function?

File: executor/aggfuncs/func_avg.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+	baseAvgFloat64
+}
+
+func (e *avgPartial4Float64) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) error {
+	p := (*partialResult4AvgFloat64)(pr)
+	for _, row := range rowsInGroup {
+		inputSum, isNull, err := e.args[1].EvalReal(sctx, row)
+		if err != nil {
+			return errors.Trace(err)
+		} else if isNull {
```
Comment (shenli) [inline executor/aggfuncs/func_avg.go]: do not need the `else`

File: executor/aggfuncs/func_avg.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+			continue
+		}
+
+		inputCount, isNull, err := e.args[0].EvalInt(sctx, row)
+		if err != nil {
+			return errors.Trace(err)
+		} else if isNull {
```
Comment (shenli) [inline executor/aggfuncs/func_avg.go]: ditto

File: executor/aggfuncs/aggfuncs.go (hunk 1 above)
Comment (zz-jason) [reply executor/aggfuncs/aggfuncs.go]: `avg4Decimal` maybe confusing with `avgPartial4Decimal`, I prefer the old name.

**Comment by shenli** [comment]
LGTM

File: executor/aggfuncs/builder.go [hunk 7]
```diff
@@ -58,6 +59,42 @@ func buildSum(aggFuncDesc *aggregation.AggFuncDesc, ordinal int) AggFunc {
 
 // buildCount builds the AggFunc implementation for function "AVG".
```
Comment (Xuanwo) [inline executor/aggfuncs/builder.go]: `buildCount` here may be a type?

**Pull request merged**